### PR TITLE
RMA: zcopy support

### PR DESCRIPTION
**Pull request by alex-mikheev**
initial commit
@yosefe please take a look

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2963/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/482/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2428/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2964/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/483/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2429/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2431/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2966/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/485/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2432/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2434/ for details (Mellanox internal link).

File: src/ucp/rma/basic_rma.c [hunk 1]
```diff
@@ -30,6 +30,18 @@ typedef ucs_status_t
         return UCS_ERR_INVALID_PARAM; \
     }
 
+static UCS_F_ALWAYS_INLINE int 
+ucp_rma_put_is_zcopy(ucp_ep_rma_config_t *rma_config, size_t length)
+{
+    return length > rma_config->put_zcopy_thresh;
```
Comment (yosefe) [inline src/ucp/rma/basic_rma.c]: \>=

File: src/ucp/rma/basic_rma.c [hunk 2]
```diff
@@ -30,6 +30,18 @@ typedef ucs_status_t
...
+}
+
+static UCS_F_ALWAYS_INLINE int 
+ucp_rma_get_is_zcopy(ucp_ep_rma_config_t *rma_config, size_t length)
+{
+    return length > rma_config->get_zcopy_thresh;
```
Comment (yosefe) [inline src/ucp/rma/basic_rma.c]: \>=

File: src/ucp/rma/basic_rma.c [hunk 3]
```diff
@@ -60,19 +76,38 @@ ucp_rma_request_advance(ucp_request_t *req, size_t frag_length,
     }
 }
 
+/**
```
Comment (yosefe) [inline src/ucp/rma/basic_rma.c]: this comment should go inside the function:
function comment -> describe function API
internal code comment -> describe implementation

File: src/ucp/rma/basic_rma.c [hunk 4]
```diff
@@ -83,16 +118,49 @@ ucp_rma_request_init(ucp_request_t *req, ucp_ep_h ep, const void *buffer,
...
     req->send.uct_comp.count  = 0; 
-    req->send.uct_comp.func   = ucp_rma_request_bcopy_completion;
-
 #if ENABLE_ASSERT
     req->send.cb              = NULL;
 #endif
+    if (zcopy) {
+        req->send.uct_comp.func = ucp_rma_request_zcopy_completion;
+        return ucp_request_send_buffer_reg(req, lane);
+    }
```
Comment (yosefe) [inline src/ucp/rma/basic_rma.c]: IMHO would be nicer to have 'else' here

File: src/ucp/rma/basic_rma.c [hunk 5]
```diff
@@ -117,29 +185,65 @@ ucp_progress_put_inner(ucp_request_t *req, uct_rkey_t uct_rkey,
...
-    frag_length = ucs_min(rma_config->max_get_bcopy, req->send.length);
-    status = UCS_PROFILE_CALL(uct_ep_get_bcopy,
-                              ep->uct_eps[req->send.lane],
-                              (uct_unpack_callback_t)memcpy,
-                              (void*)req->send.buffer,
-                              frag_length,
-                              req->send.rma.remote_addr,
-                              uct_rkey,
-                              &req->send.uct_comp);
+    if (zcopy) {
```
Comment (yosefe) [inline src/ucp/rma/basic_rma.c]: maybe better to have !zcopy as first case - for performance, and consistency with put

File: src/ucp/rma/basic_rma.c (hunk 3 above)
Comment (alex-mikheev) [inline src/ucp/rma/basic_rma.c]: the comment is relevant for the two function (bcopy/zcopy_completion) and it is also somewhat relevant for the rma_request_adcance. May be I should merge the comment with the one before ucp_rma_request_advance ?

File: src/ucp/rma/basic_rma.c (hunk 4 above)
Comment (alex-mikheev) [inline src/ucp/rma/basic_rma.c]: I will mark zcopy code path as unlikely()

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2990/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/519/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2455/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/522/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2456/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2461/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@shamisp pls take a look

**Comment by shamisp** [comment]
👍

**Pull request merged**